### PR TITLE
Prebid 7: auction-specific first party data and refactoring of FPD internals

### PR DESCRIPTION
**Pull request by dgirardi**
## Type of change

- [x] Feature
- [x] Refactoring (no functional changes, no api changes)

## Description of change
This restructures the way Prebid handles first party data to enable auction-specific FPD (https://github.com/prebid/Prebid.js/issues/7651)

﻿The changes are:

 - `pbjs.requestBids` now accepts an `ortb2` parameter, which will get merged into (and overrides) the global `ortb2` config only for that particular auction
 - `requestBids` now "freezes" FPD for a particular auction into global (`getConfig('ortb2')` + the auction-specific `ortb2`) and bidder fragments that get passed down to a new hook, `startAuction` (and further down the stack to auctionManager, adapterManager, etc)
 - FPD modules and RTD modules have been refactored to hook into `startAuction` (instead of `requestBids`), and they read from and modify its FPD parameters instead of calling `getConfig` / `setConfig`
 - bid requests and bidd*er* requests (the parameters passed to bidders' `buildRequests`) are now augmented with an `ortb2` property that replaces the use of `getConfig('ortb2')` for bid adapters. All adapters have been refactored to read from those instead of calling `getConfig`.
 - FPD submodules (currently just enrichment and validation) now run on every auction, and the global `refreshFpd` has been removed - with FPD now being potentially different for each one, I don't think the previous run-once approach makes sense. That said, both submodules can be improved to memoize some of their processing - that's not included in this PR.

The choice to put `ortb2` into every bid request is due to some adapters doing their FPD logic in the validation phase (`isBidRequestValid`) or in `transformBidParameters`, where the bidd*er* request is not available.

Note that two RTD providers - weborama and jwPlayer - were already populating `bid.ortb2` under some circumstances. Most adapters were blind to it (I know only of proxistore looking for it - related is also https://github.com/prebid/Prebid.js/pull/8320). I have refactored those two to update the bidder-level ortb2 config instead.

## Other information

Docs PR TBD

### CODE REVIEW

**Comment by dgirardi** [comment]
Tagging:

 - @desbo &  @dreischer for changes to the permutive RTD adapter
 - @karimMourra  for jwPlayer
 - @peczenyj for weborama

**Comment by peczenyj** [comment]
LGTM

I am currently working on this module, I will wait to rebase it properly

**Review by karimMourra — APPROVED** [review]

File: modules/jwplayerRtdProvider.js [hunk 1]
```diff
@@ -158,7 +158,7 @@ export function enrichAdUnits(adUnits) {
       const contentId = getContentId(vat.mediaID);
       const contentData = getContentData(vat.segments);
       const targeting = formatTargetingResponse(vat);
-      enrichBids(adUnit.bids, targeting, contentId, contentData);
+      enrichBids(ortb2Fragments?.bidder || {}, adUnit.bids, targeting, contentId, contentData);
```
[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/jwplayerRtdProvider.js:161]: true, this or the default param value is superfluous - I missed it

**Comment by dgirardi** [comment]
@peczenyj this is a breaking change for version 7; you'll probably want to fix for 6 independently, using the current `getBidderConfig` / merge / `setBidderConfig` strategy.

**Comment by peczenyj** [comment]
> @peczenyj this is a breaking change for version 7; you'll probably want to fix for 6 independently, using the current `getBidderConfig` / merge / `setBidderConfig` strategy.

Well, I try to structure this refactor to be easier to rebase  https://github.com/prebid/Prebid.js/pull/8350

**Review by desbo — APPROVED** [review]

File: src/config.js [hunk 2]
```diff
@@ -314,42 +314,52 @@ export function newConfig() {
...
+  function _getRestrictedConfig() {
+    // This causes reading 'ortb2' to throw an error; with prebid 7, that will almost
+    // always be the incorrect way to access FPD configuration (https://github.com/prebid/Prebid.js/issues/7651)
+    // code that needs the ortb2 config should explicitly use `getAnyConfig`
+    // TODO: this is meant as a temporary tripwire to catch inadvertent use of `getConfig('ortb')` as we transition.
+    // It should be removed once the risk of that happening is low enough.
+    const conf = _getConfig();
+    Object.defineProperty(conf, 'ortb2', {
+      get: function () {
+        throw new Error('invalid access to \'orbt2\' config - use request parameters instead');
```
Comment (patmmccann) [inline src/config.js:326]: this is applicable to bidders, but what if publishers what to access the config, for example, to achieve #6240

Comment (dgirardi) [reply src/config.js:326]: the publisher API does not hit this code path - the global `pbjs.getConfig` and `pbjs.readConfig` are aliased to `getAnyConfig` and `readAnyConfig`.

**Review by patmmccann — APPROVED** [review]

**Review by mmoschovas — APPROVED** [review]

**Pull request merged**